crowbar_framework/lib/crowbar/checks.rb

```diff
@@ -0,0 +1,170 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "resolv"
+require "ipaddr"
+require "socket"
+require "logger"
+
+module Crowbar
+  class NetworkChecks
+
+    attr_accessor :logger
+
+    class << self
+      def ip_addrs(fqdn, version = "ipv4")
+        ip_addrs = []
+
+        Resolv.getaddresses(fqdn).each do |address|
+          ip_addr = IPAddr.new(address)
+          if version == "ipv6"
+            ip_addrs.push address if ip_addr.ipv6?
+          else
+            ip_addrs.push address if ip_addr.ipv4?
+          end
+        end
+
+        ip_addrs
+      end
+    end
+
+    def initialize(logger = ::Logger.new(STDOUT))
+      self.logger = logger
+    end
+
+    def run_all_checks
+      [
+        :fqdn_detected?,
+        :ip_address_resolved?,
+        :loopback_resolved?,
+        :ip_configured?,
+        :network_json_valid?,
+        :firewall_disabled?,
+        :ping_succeeds?
+      ].each do |check|
+        return false unless send(check)
+      end
+
+      true
+    end
+
+    def fqdn_detected?
+      if hostname.nil? || domain.nil?
+        logger.error("Unable to detect fully-qualified hostname.")
+        return false
+      end
+
+      true
+    end
+
+    def ip_address_resolved?
+      if ipv4_addrs.empty? && ipv6_addrs.empty?
+        logger.error("Could not resolve #{fqdn} to an IPv4 or IPv6 address.")
+        return false
+      end
+
+      true
+    end
+
+    def loopback_resolved?
+      if ipv4_addrs.find { |e| /^127/ =~ e } || ipv6_addrs.find { |e| /^::[0-9]$/ =~ e }
+        logger.error("#{fqdn} resolves to a loopback address.")
+        return false
+      end
+
+      true
+    end
+
+    def ip_configured?
+      ipv4_configured = false
+      ipv6_configured = false
+
+      addr_infos = Socket.ip_address_list
+      addr_infos.each do |addr_info|
+        ipv4_configured = true if ipv4_addrs.include?(addr_info.ip_address)
+        ipv6_configured = true if ipv6_addrs.include?(addr_info.ip_address)
+      end
+
+      unless ipv4_configured || ipv4_addrs.empty?
+        logger.error("No local interface is configured with one of these addresses #{ipv4_addrs.join(" ")}.")
+        return false
+      end
+
+      unless ipv6_configured || ipv6_addrs.empty?
+        logger.error("No local interface is configured with one of these addresses #{ipv6_addrs.join(" ")}.")
+        return false
+      end
+
+      true
+    end
+
+    def network_json_valid?
+      if File.exist?("/etc/crowbar/network.json")
+        network_json = "/etc/crowbar/network.json"
+      else
+        network_json = "/opt/dell/chef/data_bags/crowbar/template-network.json"
+      end
+
+      unless system("network-json-validator --admin-ip \"#{ipv4_addrs.first}\" #{network_json}")
```

Comment:
Not a new issue, but more visible here: the fact that we rely on the fact that the address we care about is first is kind of wrong...